### PR TITLE
i#5843 scheduler: Add direct thread switch support

### PR DESCRIPTION
**Pull request by derekbruening**
Adds support for the TRACE_MARKER_TYPE_DIRECT_THREAD_SWITCH marker, when it appears after TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL.  The scheduler directly switches to the target thread if it is on the ready queue.  Performing a forced migration if the target is running on another output is not yet implemented.  Once i/o wait states are added, waking up a target thread will be added, but that is future work as well.

Adds a DEPENDENCY_DIRECT_SWITCH_BITFIELD and renamed DEPENDENCY_TIMESTAMPS
to DEPENDENCY_TIMESTAMP_BITFIELD so we can combine them, and made a new
enum entry DEPENDENCY_TIMESTAMPS which combines the two bitfields, which is
what nearly every use case should want while still giving us control and
without really breaking compatibility (and by providing bits and
combinations the enum type is all that's needed still).

Adds a unit test where the schedule would clearly be different without the switch target.

Issue: #5843

### CODE REVIEW

**Comment by derekbruening** [comment]
ub22 x32 failure is off.simple #6416
win64 failures: replaceall #5412, api.dis/decenc #6411
win32 failures: floatpc #2267

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 1]
```diff
@@ -2964,6 +2964,100 @@ test_inactive()
...
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_A);
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs_B)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_B);
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs_C)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_C);
+    // The lockstep treats markers like instructions so we have a long sequence
+    // of dots before we do the direct swap to C and back.
+    // TODO i#5843: Once we have i/o wait times we can make this more interesting
+    // by adding a D and having C be unschedule-able at switch time.
+    static const char *const CORE0_SCHED_STRING = "..AA........CC......A.C...";
```
Comment (abhinav92003) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:3038]: This schedule would be followed also if we had specified DEPENDENCY_TIMESTAMP right? But the test is verifying that it is followed even though we did not specify it?

Comment (abhinav92003) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:3038]: What happens in DEPENDENCY_TIMESTAMP if the trace has timestamp+cpu markers that contradict the switchto targets. That shouldn't happen right?

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/scheduler/scheduler.h [hunk 2]
```diff
@@ -945,8 +945,9 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
         bool order_by_timestamp = false;
         // Global ready queue counter used to provide FIFO for same-priority inputs.
         uint64_t queue_counter = 0;
-        // Used to switch on the insruction *after* a blocking syscall.
+        // Used to switch on the instruction *after* a blocking syscall.
         bool processing_blocking_syscall = false;
+        input_ordinal_t switch_to_input = INVALID_INPUT_ORDINAL;
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:950]: I figured I would keep it generic as we don't know the final name of the targeted switch feature once it successfully makes its way into the mainline kernel.

[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:950]: Added a comment better explaining the field.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 3]
```diff
@@ -2964,6 +2964,100 @@ test_inactive()
...
+    static constexpr memref_tid_t TID_BASE = 100;
+    static constexpr memref_tid_t TID_A = TID_BASE + 0;
+    static constexpr memref_tid_t TID_B = TID_BASE + 1;
+    static constexpr memref_tid_t TID_C = TID_BASE + 2;
+    std::vector<trace_entry_t> refs_A = {
+        make_thread(TID_A),
+        make_pid(1),
+        make_timestamp(101),
+        make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+        make_instr(/*pc=*/101),
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:2982]: Added an extra zero to the timestamps.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:3038]: This is a good point.  I added a DEPENDENCY_DIRECT_SWITCH_BITFIELD and renamed DEPENDENCY_TIMESTAMPS to DEPENDENCY_TIMESTAMP_BITFIELD so we can combine them, and made a new enum entry DEPENDENCY_TIMESTAMPS which combines the two bitfields, which is what nearly every use case should want while still giving us control and without really breaking compatibility (and by providing bits and combinations the enum type is all that's needed still).  I changed this test to use the direct switch one.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 4]
```diff
@@ -2964,6 +2964,100 @@ test_inactive()
...
+    // of dots before we do the direct swap to C and back.
+    // TODO i#5843: Once we have i/o wait times we can make this more interesting
+    // by adding a D and having C be unschedule-able at switch time.
+    static const char *const CORE0_SCHED_STRING = "..AA........CC......A.C...";
+    static const char *const CORE1_SCHED_STRING = "..BBBBBBBBBBBBBBBBBBBB.";
+
+    std::vector<scheduler_t::input_workload_t> sched_inputs;
+    sched_inputs.emplace_back(std::move(readers));
+    scheduler_t::scheduler_options_t sched_ops(scheduler_t::MAP_TO_ANY_OUTPUT,
+                                               scheduler_t::DEPENDENCY_IGNORE,
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:3044]: This is a good point. I added a DEPENDENCY_DIRECT_SWITCH_BITFIELD and renamed DEPENDENCY_TIMESTAMPS to DEPENDENCY_TIMESTAMP_BITFIELD so we can combine them, and made a new enum entry DEPENDENCY_TIMESTAMPS which combines the two bitfields, which is what nearly every use case should want while still giving us control and without really breaking compatibility (and by providing bits and combinations the enum type is all that's needed still). I changed this test to use the direct switch one.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:3038]: Do you mean A does a switch to C, and A's timestamp pre-syscall is later than C's timestamp?  The schedule will go ahead with the switch.  The timestamps for MAP_TO_ANY_OUTPUT are only used to order the ready queue.  I would agree it shouldn't happen in a real trace.  Maybe a candidate for an invariant check?  I added a comment about possibly adding such a check.

**Comment by derekbruening** [comment]
Wow, no flaky failures!

**Pull request merged**